jekyll/_cci2/optimization-cookbook.md

```diff
@@ -0,0 +1,210 @@
+
+Let's take a closer look at the testing process in the figure above to better understand the time it took to complete the tests.
+
+The following steps were performed during testing:
+
+1) The build job created a Docker image, which contained only runtime dependencies. 
+2) The build job dumped the image to a file with `docker save`, and then persisted it in the workspace. 
+3) Two tests jobs were run to restore the base image from the workspace.
+4) The test jobs built on this base image to create an image with all the extra modules required to run the tests. 
+5) The test jobs started dependencies, and the tests were finally initiated.
+
+Typically, performing setup once, and then performing `fan out` steps, is a traditional way to reduce resource usage; however, in this example, the `fan out` steps proved to be very expensive in the following ways:
+
+- Issuing `docker save` to dump the built image to a file took around **30** seconds.
+- Persisting the image to the workspace added another **60** seconds.
+- The test jobs then had to attach the workspace and load the base image, adding another **30** seconds to the testing.
+- The test jobs started dependencies (Redis, Cassandra, and PostgreSQL) with `docker-compose`, which required the use of the machine executor. This added an additional **30-60** second startup time compared to the docker executors.
+- Because the base image from the build job contained only runtime dependencies, a docker image had to be built, extending the base to add dependencies for testing. This added  about **70** seconds.
+
+As you can see, there is a a significant amount of time being spent setting up the tests, without any actual tests being performed. In fact, this approach required 6.5 minutes before the actual tests were run, which took another 6.5 minutes.
+
+### Test Preparation Optimization
+
+Knowing that ~13 minutes was too long to perform the steps in this workflow, the following approaches were taken to optimize this workflow to reduce the time it took to run this workflow.
+
+#### Changing the CI Test Workflow
+
+The CI test workflow was changed to no longer depend on building the base image. The test jobs were also changed to lauch auxiliary services using CircleCI's docker executor native service container support instead of using `docker-compose`. Finally,`tox` was run from the main container to install dependencies and run tests, which eliminates minutes used to save and then restore the image from the workspace. This also eliminated the extra start-up costs of the machine executor.
+
+#### Dependency Changes
+
+Installing dependencies in the primary container on CircleCI, rather than relying on a Dockerfile, may enable you to use CircleCI's caching to speed up `virtualenv` creation. Even if you do not have access to this shared cache when installing dependencies, you may save around **90** seconds when building the virtualenvs.
```

Comment:
> Even if you do not have access to this shared cache when installing dependencies, you may save around **90** seconds when building the virtualenvs.

I'm afraid I don't follow. How can you save 90 seconds when building the virtualenvs "even if you do not have access to the shared cache"?

Reply:
@stig You're right, this was confusing, so I just went ahead and removed this from the page. Thanks for the good catch.